source/agora/consensus/PreImage.d

```diff
@@ -291,17 +291,71 @@ public struct PreImageCycle
 
     public static immutable ulong NumberOfCycles = 100;
 
-    /// Construct an instance with the provided cycle length and number of cycles
+    /***************************************************************************
+
+        Construct an instance based on a secret
+
+        The secret will be hashed in a predictable manner, and this will give
+        the 'seed' hash, which will correspond to the last pre-image the node
+        will be able to reveal (at height `cycles * cycle_length`).
+
+        This overload is provided for convenience. To have more control over
+        the pre-image, see the overload accepting a Hash.
+
+        Params:
+          secret = The secret to use to generate the pre-image
+          cycle_length = The length of a cycle
+          cycles = Number of cycles to generate (defaults to 100)
+          initial_seek = The value to seek at after generating the seeds
+
+    ***************************************************************************/
+
     public this (in Scalar secret, in uint cycle_length, in ulong cycles = PreImageCycle.NumberOfCycles,
-        Height initial_seek = Height(0))
+        in Height initial_seek = Height(0))
+    {
+        // Using this predictable scheme ensures that the pre-image can
+        // be recovered in case of a crash or if another node takes over
+        // (e.g. disk failure from a server leads to a backup server starting
+        // with no state or way to recover it).
+        const cycle_seed = hashMulti(secret, "consensus.preimages", 0);
```

Comment:
Do we need to keep this `nonce` of zero still?

Reply:
Not that I can think of. The only upside is to be consistent with how other parameters are computed.

Reply:
It's not necessary. I will remove it which should have been removed in the recent commit I have submitted.